### PR TITLE
Added support for parameter strong types #1083

### PR DESCRIPTION
**Pull request by BernieWhite**
## PR Summary

- New features:
  - Added support for parameter strong types. #1083
    - The value of string parameters can be tested against the expected type.
    - When configuring a location strong type, the parameter value must be a valid Azure location.
    - When configuring a resource type strong type, the parameter value must be a matching resource Id.

## PR Checklist

- [x] PR has a meaningful title
- [x] Summarized changes
- [x] Change is not breaking
- [x] This PR is ready to merge and is not **Work in Progress**
- **Rule changes**
  - [x] Unit tests created/ updated
  - [x] Rule documentation created/ updated
  - [x] Link to a filed issue
  - [x] [Change log](https://github.com/Azure/PSRule.Rules.Azure/blob/main/docs/CHANGELOG-v1.md) has been updated with change under unreleased section

### CODE REVIEW

File: src/PSRule.Rules.Azure/Data/Template/ValidationIssue.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        None = 0,
+
+        Parameter = 1
+    }
+
+    internal sealed class ValidationIssue
+    {
+        private ValidationKind kind;
+        private string name;
+        private string message;
```
Comment (ArmaanMcleod) [inline src/PSRule.Rules.Azure/Data/Template/ValidationIssue.cs:17]:
```suggestion
        private readonly ValidationKind kind;
        private readonly string name;
        private readonly string message;
```

File: docs/en/rules/Azure.Template.ParameterStrongType.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+Template string parameters can optionally specify a strong type.
+When parameter files are expanded, if the parameter value does not match the type this rule fails.
+Support is provided by PSRule for Azure for the following types:
+
+- Resource type - Specify a resource type.
+  For example `Microsoft.OperationalInsights/workspaces`.
+  If a resource type is specified the parameter value must be a resource id of that type.
+- Location - Specify `location` as the strong type.
+  If `location` is specifies, the parameter value must be a valid Azure location.
```
Comment (ArmaanMcleod) [inline docs/en/rules/Azure.Template.ParameterStrongType.md:25]:
```suggestion
  If `location` is specified, the parameter value must be a valid Azure location.
```

**Review by ArmaanMcleod — APPROVED** [review]
Looks great. I just added some small suggestions 🙂

**Comment by BernieWhite** [comment]
@ArmaanMcleod Thanks. Great suggesions.

**Pull request merged**